### ISSUE TITLE
[tutorial] composite compute kernels

### ISSUE THREAD

**Issue by palindromoroz** [halide/Halide#2054]
It is not clear to me what is the recommended way of combining multiple computation kernels into one larger operation and reuse the code. For example if I have blur and gradient implemented separately and I want to combine the code and write a new schedule for the composite operation. Maybe it would be nice to make an example tutorial for that. Main part that is unclear to me is how to deal with the Vars and the Funcs. Current larger pipe examples have them just as global in the composite class which is not very modular.

**Comment by SanderVocke** [comment]
This is what I do - not recommended by anyone, but it works well for me. 

I put the functionality in a class constructor, which takes any inputs and Vars used as arguments. Then expose all intermediate Funcs used as public class members. This way, all scheduling can happen externally. You could also include "standalone" schedules as member functions.

A small example:

Two "modular kernels":
```
class KernelA{
public:
  KernelA( Var x, Var y, Func input ) {
    //give Funcs names for IR inspection purposes
    intermediate_func = Func("intermediate_func");
    output = Func("output"); 
    
    intermediate_func(x, y) = input(x, y) * 2;
    output(x, y) = intermediate_func(x, y) *3;
  }

  Func output;
  Func intermediate_func;
}

class KernelB{
public:
  KernelB( Var x, Var y, Func input ) {
    //give Funcs names for IR inspection purposes
    intermediate_func = Func("intermediate_func");
    output = Func("output"); 
    
    intermediate_func(x, y) = input(x, y) * 5;
    output(x, y) = intermediate_func(x, y) * 6;
  }

  Func output;
  Func intermediate_func;
}
```

And two pipelines using them in different combinations and schedules:

```
Var x("x"), y("y");
Func input("input");

//variant 1
KernelA kernel_a(x, y, input);
kernel_a.intermediate_func.compute_root();
... = kernel_a.output.realize(...);

//variant 2
KernelA kernel_a(x, y, input);
KernelB kernel_b(x, y, kernel_a.output);
kernel_a.intermediate_func.compute_at(kernel_b.output, y);
... = kernel_b.output.realize(...);
```

I'm curious what others think of this approach for larger complex pipelines.

**Comment by SanderVocke** [comment]
Since the new-style Generators were introduced, the above is more or less equivalent to using multiple Generators which expose the intermediate, schedulable Funcs as members:

https://github.com/halide/Halide/wiki/Generator-Enhancements 

(bottom of the page). Probably it's better to use those than to make your own classes from scratch.

**Comment by palindromoroz** [comment]
Yes, those come close. Not sure you can use them for building kernel combinations except as libraries.

**Comment by palindromoroz** [comment]
Hey. You are right. Did not check the update of the page. Not the easiest way but flexible and supported.
